neuralop/layers/attention_kernel_integral.py

```diff
@@ -0,0 +1,212 @@
+import torch
+from torch import nn
+import math
+from torch.nn.init import xavier_uniform_, zeros_
+
+
+class AttentionKernelIntegral(torch.nn.Module):
+    """
+    Kernel integral transform with attention
+    Computes \int_{Omega} k(x, y) * f(y) dy,
+    where:
+          K(x, y) = \sum_{c=1}^d \q_c(x) * \k_c(y)
+          f(y) = v(y)
+
+    Parameters
+    ----------
+    in_channels : int, input channels
+    out_channels : int, output channels
+    n_heads : int, number of attention heads in multi-head attention
+    head_n_channels : int, dimension of each attention head, determines how many function bases to use for the kernel
+                      k(x, y) = \sum_{c=1}^d \q_c(x) * \k_c(y), head_n_channels controls the d
+    pos_dim : int, dimension of the domain, determines the dimension of coordinates
+    use_positional_encoding : bool, whether to use positional encoding to encode query and key
+    project_query : bool, whether to project the query function with pointwise linear layer
+                   (this is sometimes not needed when using cross-attention)
+    """
+
+    def __init__(self,
+                 in_channels,
+                 out_channels,
+                 n_heads,
+                 head_n_channels,
+                 pos_dim,
+                 use_positional_encoding=True,    # use positional encoding
+                 project_query=True,
+                 ):
+        super().__init__()
+        self.layers = nn.ModuleList([])
+        self.n_heads = n_heads
+        self.head_n_channels = head_n_channels
+
+        self.project_query = project_query
+        if project_query:
+            self.to_q = nn.Linear(in_channels, head_n_channels * n_heads, bias=False)
+        else:
+            self.to_q = nn.Identity()
+
+        self.to_k = nn.Linear(in_channels, head_n_channels * n_heads, bias=False)
+
+        self.k_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+        self.v_norm = nn.InstanceNorm1d(head_n_channels, affine=False)
+
+        self.to_v = nn.Linear(in_channels, head_n_channels * n_heads, bias=False)
+
+        self.to_out = nn.Linear(head_n_channels * n_heads, out_channels) \
+            if head_n_channels * n_heads != out_channels else nn.Identity()
+
+        self.use_positional_encoding = use_positional_encoding
+        self.pos_dim = pos_dim
```

Comment:
Why do we need `self.pos_dim`? We can just infer it from `pos_x` at runtime.

Reply:
Yes